go/vt/mysqlctl/mysqlshellbackupengine.go

```diff
@@ -0,0 +1,340 @@
+	// drain a tablet when taking a backup
+	mysqlShellBackupShouldDrain = false
+	// disable redo logging and double write buffer
+	mysqlShellSpeedUpRestore = false
+
+	MySQLShellPreCheckError = errors.New("MySQLShellPreCheckError")
+)
+
+// MySQLShellBackupManifest represents a backup.
+type MySQLShellBackupManifest struct {
+	// BackupManifest is an anonymous embedding of the base manifest struct.
+	// Note that the manifest itself doesn't fill the Position field, as we have
+	// no way of fetching that information from mysqlsh at the moment.
+	BackupManifest
+
+	// Location of the backup directory
+	BackupLocation string
+	// Params are the parameters that backup was created with
+	Params string
+}
+
+func init() {
+	for _, cmd := range []string{"vtcombo", "vttablet", "vtbackup", "vttestserver", "vtctldclient"} {
+		servenv.OnParseFor(cmd, registerMysqlShellBackupEngineFlags)
+	}
+}
+
+func registerMysqlShellBackupEngineFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&mysqlShellBackupLocation, "mysql_shell_backup_location", mysqlShellBackupLocation, "location where the backup will be stored")
+	fs.StringVar(&mysqlShellFlags, "mysql_shell_flags", mysqlShellFlags, "execution flags to pass to mysqlsh binary to be used during dump/load")
+	fs.StringVar(&mysqlShellDumpFlags, "mysql_shell_dump_flags", mysqlShellDumpFlags, "flags to pass to mysql shell dump utility. This should be a JSON string and will be saved in the MANIFEST")
+	fs.StringVar(&mysqlShellLoadFlags, "mysql_shell_load_flags", mysqlShellLoadFlags, "flags to pass to mysql shell load utility. This should be a JSON string")
+	fs.BoolVar(&mysqlShellBackupShouldDrain, "mysql_shell_should_drain", mysqlShellBackupShouldDrain, "decide if we should drain while taking a backup or continue to serving traffic")
+	fs.BoolVar(&mysqlShellSpeedUpRestore, "mysql_shell_speedup_restore", mysqlShellSpeedUpRestore, "speed up restore by disabling redo logging and double write buffer during the restore process")
+}
+
+// MySQLShellBackupEngine encapsulates the logic to implement the restoration
+// of a mysql-shell based backup.
+type MySQLShellBackupEngine struct {
+}
+
+const (
+	mysqlShellBackupBinaryName = "mysqlsh"
+	mysqlShellBackupEngineName = "mysqlshell"
+)
+
+func (be *MySQLShellBackupEngine) ExecuteBackup(ctx context.Context, params BackupParams, bh backupstorage.BackupHandle) (result BackupResult, finalErr error) {
+	params.Logger.Infof("Starting ExecuteBackup in %s", params.TabletAlias)
+
+	err := be.backupPreCheck()
+	if err != nil {
+		return BackupUnusable, vterrors.Wrap(err, "failed backup precheck")
+	}
+
+	start := time.Now().UTC()
+	location := path.Join(mysqlShellBackupLocation, params.Keyspace, params.Shard, start.Format("2006-01-02_15-04-05"))
+
+	args := []string{}
+
+	if mysqlShellFlags != "" {
+		args = append(args, strings.Fields(mysqlShellFlags)...)
+	}
+
+	args = append(args, "-e", fmt.Sprintf("util.dumpSchemas([\"vt_%s\"], %q, %s)",
```

Comment:
you mean escape like in MySQL as `` `vt_keyspace` ``? I am not sure MySQL Shell supports or expects it, I will verify that

Reply:
You can't assume that the database name is `vt_keyspace`. It is determined by the `--init_db_name_override` flag.

Reply:
This brings up another point. Typically when we do the regular kind of backup we also backup the sidecar db, which defaults to `_vt`. Does the mysqlshell backup specifically backup only the actual keyspace/database? What are the implications of this?

Reply:
oh, it is possible to backup multiple db/schemas, we just need to specify them. is this the right place to get the name of the sidecar db?
https://github.com/vitessio/vitess/blob/c51ee4b7e3da381843ca6ae37fa404715bf8d459/go/constants/sidecar/name.go#L23-L25

also, I am curious what are the consequences if `_vt` is not in place, we tried some of these restores without any issues but maybe we are losing some metadata?

Reply:
moved the code to call `sidecar.GetName()`

Reply:
Re the sidecar db, you may not see any bad effects unless and until you try to promote the new replica, or unless you are in the midst of a reshard / online DDL etc. Also schema tracking. If any of these features are active, the replication stream could eventually break after the replica is up.

Reply:
@rvrangel it seems like `vt_` + keyspace name is still hardcoded. That needs to be fixed. You will need to somehow look that up and thread it through.

Reply:
@rvrangel on that note, something I only realized when discussing your PR with the team, are you taking backups of explicit databases? Is there a way in `mysqlshell` to abckup all databases? I'd assume the `mysql` database needs backing up?

Reply:
that's a good question. we can backup all databases (that's the difference between `dumpInstance()` vs `dumpSchemas()` if I remember correctly), I guess for our setup we create mysql grants every time we provision the tablets in a different way, is that why you are thinking of backing up the `mysql` DB too?

we could just add it to the list if other databases are supposed to be ephemeral from the Vitess point of view, otherwise would backing up all of them be preferable?

Reply:
Vitess-wise we should backup all databases. When vitess restores MySQL from backup, it makes no assumptions on existing data, nor does it populate any specific data prior/after backup. It expects the backup to just work and restore MySQL correctly.

Reply:
okay, I have updated it to use `dumpInstance()` instead, which should include all the databases now.

Reply:
@shlomi-noach one interesting side effect of this I was thinking when [discussing a related issue](https://github.com/planetscale/vitess-operator/pull/586#issuecomment-2324490828) on this other PR, is that we now dump all databases, but we only drop the ones managed by vitess on restore and that could causes during restore since we only drop the ones managed by vitess:
https://github.com/vitessio/vitess/pull/16295/files#diff-7445e78fc6fe2a32a3e8b6a01f6159a07e3b79d29f83eb81b22f3601a1e91436R275-R277

what do you believe is the best course of action for vitess to take here? leave it as is (so the user will need to clear MySQL before any restore that already has data on disk if they are using other databases) or then make sure we are dropping all databases before the restore takes place?

Reply:
Right. So this is something completely new to the vitess backup/restore process, and I did not catch that earlier.

With both existing backup engines (`builtin` and `xtrabackup`) Vitess assumes the entire schema, including `mysql.*` tables, is restored from backup in one shot. This also aligns with GTID positioning as `gtid_executed` is just restored as part of the entire table set in `mysql.gtid_executed`.

So what I'm missing right now is how `mysqlshell` backs up and restores. From a [quick read](https://www.percona.com/blog/backup-and-restore-using-mysql-shell/) I gather `mysqlshell` doesn't actually backup` the `mysql` database? Or can it be forced to do so?

And, upon restore, am I correct to understand there needs to be a database up and running, and with some appropriate user privileges, before issuing a restore? What are the minimal requirements?

The most vitess-compliant method would be to issue a complete dump and complete restore. Is there a strong argument against doing so, and for only backing up a subset of databases? I don't see how that would play nicely with GTID, but if there's a known solution, please let me know.

Reply:
includeSchemas, `dumpInstance()` [doesn't support backing up these even if using](https://dev.mysql.com/doc/mysql-shell/8.0/en/mysql-shell-utilities-dump-instance-schema.html#mysql-shell-utilities-dump-opt-filtering) `includeSchemas`:

>  includeSchemas: array of strings
>
>    (Instance dump utility only) Include only the named schemas in the dump. This option is available from MySQL Shell 8.0.28. You cannot include the information_schema, mysql, ndbinfo, performance_schema, or sys schemas by naming them on this option. If you want to dump one or more of these schemas, you could do this using the schema dump utility util.dumpSchemas()

I think this is less of an issue for us as we deal with MySQL grants outside of Vitess, and not every table will be replicated, but @ejortegau and I were brainstorming on it and one possible solution if you expect grants to be handled on the mysql database is to use `loadUsers` option in MySQL Shell. MySQL Shell already includes users by default on the backups with `dumpInstance()`, so we could just restore the users as they were when the backup is taken, but that requires us deleting all the users currently on MySQL (except the current user) otherwise it will error out if they already exist, [see the docs](https://dev.mysql.com/doc/mysql-shell/8.0/en/mysql-shell-utilities-load-dump.html#mysql-shell-utilities-load-dump-opt-filtering):

> loadUsers: [ true | false ]
>
>    Import (true) or do not import (false) users and their roles and grants into the target MySQL instance. The default is false, so users are not imported by default. Statements for the current user are skipped. From MySQL Shell 8.0.22, if a user already exists in the target MySQL instance, an error is returned and the user's grants from the dump files are not applied. From MySQL Shell 8.0.22, you can use the excludeUsers or includeUsers option in the dump loading utility to specify user accounts to be excluded or included in the import. 

Let me know if you think that would ease the concerns with the GTID breaking under this. as mentioned above, we manage grants outside of vitess and we avoid replication with `sql_log_bin=0` when making changes there, so we haven't seen an issue with it so far

Reply:
I am answering to https://github.com/planetscale/vitess-operator/pull/586#issuecomment-2324490828 here so everything is in the same place.

> could you provide a bit more of detail of how it is re-created? when you say the data is restored from disk, do you mean you restore previous snapshots of the data (so tables and schema would already exist on disk) while starting the vttablet with --restore-from-backup?

That's correct.

The cluster is running, we then teardown the vttablets, keeping the vtdataroot and all associated data, which is what happen by default when you remove the tablets on kubernetes. We then bring the tablets back up again, and they start with the vtdataroot already filled, meaning the data will be loaded from what already exists there. Moreover, because we defined a backup configuration on our K8S yaml file, we will start vttablet with the `--restore_from_backup` flag. Which will try to load any existing backup after the tablets are up and running.


> if so, the main issue [is in here](https://github.com/vitessio/vitess/pull/16295/files#diff-7445e78fc6fe2a32a3e8b6a01f6159a07e3b79d29f83eb81b22f3601a1e91436R251-R283), we only drop those databases for receiving a RestoreFromBackup() API call, not when starting with --restore-from-backup.

Would it be possible to drop those databases always when using the mysqlsh engine?

> from the code comments, it seems if you are starting a tablet with existing data, [it needs to be wiped out before you start the ](https://github.com/slackhq/vitess/blob/main/go/vt/mysqlctl/backupengine.go?rgh-link-date=2024-09-02T11%3A21%3A35Z#L113-L116)vttablet or you need to start the restore via an API call instead of relying on --restore-from-backup.

I don't think wiping out the data manually in vtop would be a good idea, we could explore the idea of not starting vttablet with the restore from backup flag and instead do an API call when the tablet is healthy, but I would like to know if we could simplify it by simply dropping the databases always when mysqlsh engine is used.

Reply:
it would be possible, but it seems to me it would be an exception to handling `DeleteBeforeRestore`, where its value is false but we go and delete files anyway.

looking at the other backup engines, they delete current files without caring for the value of `DeleteBeforeRestore`. I can't find where this is actually used rather than just passed down all the way to the backup engine, but ignored on the other two. is there something I am missing? if we have to delete the databases anyway, I would prefer to just get rid of `DeleteBeforeRestore` if not being used, rather than having the backup engines ignore the intended value.

Reply:
This looks to be the only place where the value of `DeleteBeforeRestore ` is checked:

https://github.com/vitessio/vitess/blob/44e96c2291a2b7b4fddbd7162d2214ba58661810/go/vt/mysqlctl/backup.go#L311-L314

And it looks to me like it is senseless. I think the variable can be removed.

Reply:
@shlomi-noach would it make sense to do it in a separate PR though, so it is easier to revert if needed? I would hate for us to have to revert this PR once merged in case it causes any side effects

Reply:
`DeleteBeforeRestore` aside, would you all be comfortable with the idea adding a flag that lets makes us run `DELETE USER` on all users except the current one and than letting the user define `loadUsers` on the mysql shell restore flags on their own so these are reloaded as appropriate if they rely on replication for propagating the grants?

Reply:
> I would prefer to just get rid of DeleteBeforeRestore if not being used, rather than having the backup engines ignore the intended value.

I interpreted that in the sense that you prefer to remove the variable in this PR. It's fine if we did that on another PR.

Reply:
> would it make sense to do it in a separate PR though
Only thing to take into account is that if we do it in a separate PR, after merging this current PR the new engine won't work on vitess-operator (the restore part only). I don't mind waiting for the second PR to be ready/merged in order to merge the vitess-operator one.

Reply:
okay, lets do it in this PR then to avoid the admin work of having it separate and putting up notices, I will push a change later today